redgenes/tests/test_utils.py

```diff
@@ -0,0 +1,191 @@
+import subprocess
+from pathlib import Path
+from unittest import TestCase, main
+from unittest.mock import patch, mock_open
+from redgenes.exceptions import InvalidFna
+from redgenes.utils import (
+    generate_insert_stmt,
+    read_gff_file,
+    extract_qualifier,
+    process_qualifiers,
+    run_unzip_fna,
+    run_zip_fna,
+    copy_and_unzip,
+    run_command,
+)
+
+
+class TestUtils(TestCase):
+    def test_generate_insert_stmt(self):
+        table_name = "test_table"
+        column_names = ["col1", "col2", "col3"]
+        expected = "insert into test_table (col1, col2, col3) values (?, ?, ?);"
+        result = generate_insert_stmt(table_name, column_names)
+        self.assertEqual(result, expected)
+
+    def test_read_gff_file_file_not_found(self):
+        with self.assertRaises(FileNotFoundError):
+            read_gff_file("nonexistent.gff")
+
+    def test_extract_qualifier(self):
+        # Test case when the qualifier key exists and is a list
+        qualifiers_dict = {"key1": ["value1", "value2"], "key2": "value3"}
+        result = extract_qualifier(qualifiers_dict, "key1", str)
+        assert (
+            result == "value1"
+        ), "Failed to extract first value from a list of qualifiers"
+
+        # Test case when the qualifier key exists and is not a list
+        result = extract_qualifier(qualifiers_dict, "key2", str)
+        assert (
+            result == "value3"
+        ), "Failed to extract value when qualifier is not a list"
+
+        # Test case when the qualifier key does not exist
+        result = extract_qualifier(qualifiers_dict, "key3", int)
+        assert result is None, "Failed to return None for non-existent key"
+
+        # Test case when the qualifier value is 'None' (string)
+        qualifiers_dict = {"key1": "None"}
+        result = extract_qualifier(qualifiers_dict, "key1", str)
+        assert result is None, "Failed to return None for 'None' string value"
+
+        # Test case for converting type
+        qualifiers_dict = {"key1": "123"}
+        result = extract_qualifier(qualifiers_dict, "key1", int)
+        assert result == 123, "Failed to convert string to integer"
+
+    def test_process_qualifiers(self):
+        # Setup
+        qualifiers_dict = {
+            "key1": "value1",
+            "key2": ["value2a", "value2b"],
+            "key3": "123",
+            "key4": "None",
+        }
+        dtype_map = {
+            "key1": str,
+            "key2": str,
+            "key3": int,
+            "key4": str,
+            "key5": float,  # key5 does not exist in qualifiers_dict
+        }
+
+        # Expected result should handle the conversion and default values correctly
+        expected_result = ["value1", "value2a", 123, None, None]
+
+        # Call the function
+        result = process_qualifiers(qualifiers_dict, dtype_map)
+
+        # Assertions
+        assert (
```

Comment:
If you want, `self.assertEqual` works here. Developer preference on providing a message if it fails though many projects in our ecosystem tend to not do that